druntime/src/core/stdc/config.d

```diff
@@ -13,6 +13,35 @@
 
 module core.stdc.config;
 
+/**
+ * The choices made by each C implementation about the sizes of the
+ * fundamental types are known as "data model".
+ *
+ * See_Also: https://en.cppreference.com/w/c/language/arithmetic_types
+ */
+enum DataModel
+{
+    ILP32,  /// or 4/4/4 (int, long, and pointer are 32-bit)
+    LP64,   /// or 4/8/8 (int is 32-bit, long and pointer are 64-bit)
+    LLP64,  /// or 4/4/8 (int and long are 32-bit, pointer is 64-bit)
+}
+
+version (D_LP64)
+{
+    version (Cygwin)
+        enum dataModel = DataModel.LP64; ///
+    else version (Windows)
+        enum dataModel = DataModel.LLP64; ///
+    else
+        enum dataModel = DataModel.LP64; ///
+}
+else // 32-bit pointers
+{
+    enum dataModel = DataModel.ILP32; ///
+}
+
+static assert(__traits(compiles, typeof(dataModel)));
```

Comment:
```suggestion
    else version(Posix)
        enum dataModel = DataModel.LP64; ///
    else version(Darwin)
        enum dataModel = DataModel.LP64; ///
    else
        static assert(0, "unknown data model");
}
else // 32-bit pointers
{
    enum dataModel = DataModel.ILP32; ///
}
```

The current assert trivially holds because you default to defining either ILP32 or LP64. If someone would start porting druntime to Unicos on Cray (going by the linked cppreference), they would need to add ILP64, but this code will default set the dataModel to LP64.

Reply:
I intuitively feel that this change is not necessary. It is something of an “early optimization” in my opinion.

When anyone will port it to the Cray then we’ll add it.

Reply:
Ah, probably, I understand your point!

 I added `static assert` here not for replace default branch, but just to ensure I'm not made a typo in `dataModel` name - I can't test all branches

Reply:
druntime's platform bindings are designed to be a simple discrimination of each version:

```D
version(linux)
{
    // definitions for the Linux operating system
}
version(Windows)
{
    // definitions for the Windows operating system
}
version(OSX)
{
    // definitions for the Mac operating system
}
else
   static assert(0, "unsupported operating system");
```

They are not supposed to be deduplicated or minimized:
```D
version(Windows)
{
    // definitions for the Windows operating system
}
else
{
    // definitions for the... "not-Windows" operating system?
}
```

I wouldn't call being comprehensive about what platforms you support an "optimization". it seems like you're referring to YAGNI: You Ain't Gonna Need It. However, people are porting Druntime to new platforms all the time, so I think it's best to make that process straightforward and not bug-prone.

Reply:
_Other models are very rare. For example, ILP64 (8/8/8: int, long, and pointer are 64-bit) only appeared in some early 64-bit Unix systems (e.g. Unicos on Cray)._

By "very rare" [these guys](https://en.cppreference.com/w/c/language/arithmetic_types) mean real rarities. All(!!) other systems fit into proposed PR.

There is no need to complicate things - we can use this even without knowing at all what kind of OS is used (personally, I’m interested in bare metal) - just like the C language does

Reply:
Here is a good PR that will alerts us if implementation for Cray will be inconsistent with C: https://github.com/dlang/dmd/pull/16571

Reply:
> Here is a good PR that will alerts us if implementation for Cray will be inconsistent with C: #16571

This PR only finds inconsistencies between ImportC and druntime. For basic types both could contain the same bug. Maybe it would be a good idea, to add a test to runnable_cxx, which compares the sizes of basic types between druntime and the real C/C++ compiler.

Reply:
> Maybe it would be a good idea, to add a test to runnable_cxx, which compares the sizes of basic types between druntime and the real C/C++ compiler.

Yes, such type checks for `c_long` / `c_ulong` and sets of `int_fast*_t` / `int_least*_t` wold be nice